File: HWs/RLS-Activity/scripts/r/ejercicio2.R
# %%
library(readr)
library(ggplot2)
library(dplyr)
library(broom)

DATA_PATH <- "../../data/shampoo.csv"
PLOTS_DIR <- "../../plots/r/ejercicio2"
if (!dir.exists(PLOTS_DIR)) {
  dir.create(PLOTS_DIR, recursive = TRUE)
}

cat("Cargando:", DATA_PATH, "\n")
df <- read_csv(DATA_PATH)
cat("\nPrimeras filas:\n")
print(head(df))

ventas_col_candidates <- c("Ventas_Millones_lts", "Ventas", "ventas")
inversion_col_candidates <- c("Inversion_Millones_pesos", "Inversion", "inversion")

ventas_col <- NULL
inversion_col <- NULL

for (col in ventas_col_candidates) {
  if (col %in% names(df)) {
    ventas_col <- col
    break
  }
}

for (col in inversion_col_candidates) {
  if (col %in% names(df)) {
    inversion_col <- col
    break
  }
}

x <- as.numeric(df[[inversion_col]])
y <- as.numeric(df[[ventas_col]])

n <- length(x)
cat("\nColumnas detectadas -> y:", ventas_col, ", x:", inversion_col, "; n =", n, "\n")

# %%
cat("\n(a) Generando diagrama de dispersión y estadísticas descriptivas...\n")

p <- ggplot(data.frame(x = x, y = y), aes(x = x, y = y)) +
  geom_point() +
  labs(x = "Inversión en redes (millones de pesos)",
       y = "Ventas (millones de litros)",
       title = "Dispersión: Ventas vs. Inversión") +
  theme_minimal() +
  theme(panel.grid.minor = element_line(linetype = "dashed"))

scatter_path <- file.path(PLOTS_DIR, "scatter_plot.png")
ggsave(scatter_path, p, dpi = 300, width = 6.5, height = 4.5)
cat("Figura guardada en:", scatter_path, "\n")
print(p)

desc_data <- data.frame(
  media = c(mean(y), mean(x)),
  mediana = c(median(y), median(x)),
  desv_est = c(sd(y), sd(x)),
  min = c(min(y), min(x)),
  Q1 = c(quantile(y, 0.25), quantile(x, 0.25)),
  Q3 = c(quantile(y, 0.75), quantile(x, 0.75)),
  max = c(max(y), max(x)),
  row.names = c("Ventas (y)", "Inversión (x)")
)

cat("\nEstadísticas descriptivas:\n")
print(round(desc_data, 3))

# %%
cat("\n(b) Correlación de Pearson e intervalo de confianza al 95%...\n")
cor_result <- cor.test(x, y, method = "pearson")
r <- cor_result$estimate
pval <- cor_result$p.value

cat("r =", round(r, 4), "\n")
cat("p-valor (bilateral) =", round(pval, 6), "\n")

ci_cor <- cor_result$conf.int
cat("IC 95% para r (cor.test): [", round(ci_cor[1], 4), ",", round(ci_cor[2], 4), "]\n")

# %%
cat("\n(c) Ajustando RLS (MCO): y = beta0 + beta1*x + e ...\n")
model <- lm(y ~ x)
cat("\nParámetros estimados:\n")
print(coef(model))
cat("\nResumen del modelo:\n")
print(summary(model))

# %%
cat("\n(d) Prueba de hipótesis (one-sided, alpha=0.05)...\n")

beta1_hat <- coef(model)[2]
se_beta1 <- summary(model)$coefficients[2, 2]
df_resid <- df.residual(model)

beta1_H0 <- 0.1

t_stat <- (beta1_hat - beta1_H0) / se_beta1
p_one_sided <- 1 - pt(t_stat, df_resid)

cat("beta1_hat =", round(beta1_hat, 6), "\n")
cat("SE(beta1) =", round(se_beta1, 6), "\n")
cat("t =", round(t_stat, 4), ", df =", df_resid, "\n")
cat("p-valor (H1: beta1 > 0.1) =", round(p_one_sided, 6), "\n")

alpha <- 0.05
if (p_one_sided < alpha) {
  cat("Conclusión: Se RECHAZA H0. La evidencia sugiere un incremento > 50 mil litros por cada 500 mil pesos.\n")
} else {
  cat("Conclusión: NO se rechaza H0 al 5%. No hay evidencia suficiente para afirmar un incremento > 50 mil litros por cada 500 mil pesos.\n")
}

# %%
p_line <- ggplot(data.frame(x = x, y = y), aes(x = x, y = y)) +
  geom_point(aes(color = "Datos")) +
  geom_smooth(method = "lm", se = FALSE, aes(color = "Recta ajustada")) +
  scale_color_manual(values = c("Datos" = "black", "Recta ajustada" = "red")) +
  labs(x = "Inversión en redes (millones de pesos)",
       y = "Ventas (millones de litros)",
       title = "RLS: Ventas ~ Inversión",
       color = "Elementos") +
  theme_minimal() +
  theme(panel.grid.minor = element_line(linetype = "dashed"),
        legend.position = "bottom")

line_path <- file.path(PLOTS_DIR, "shampoo_rls_line.png")
ggsave(line_path, p_line, dpi = 300, width = 6.5, height = 4.5)
cat("Figura guardada en:", line_path, "\n")
print(p_line)

# %%

File: HWs/SLR-Activity/scripts/matlab/ejercicio2.m
%%

DATA_PATH = '../../data/shampoo.csv';
PLOTS_DIR = '../../plots/matlab/ejercicio2';
if ~exist(PLOTS_DIR, 'dir')
    mkdir(PLOTS_DIR);
end

fprintf('Cargando: %s\n', DATA_PATH);
df = readtable(DATA_PATH);
fprintf('\nPrimeras filas:\n');
disp(head(df));


ventas_col_candidates = {'Ventas_Millones_lts', 'Ventas', 'ventas'};
inversion_col_candidates = {'Inversion_Millones_pesos', 'Inversion', 'inversion'};

ventas_col = '';
inversion_col = '';
for i = 1:length(ventas_col_candidates)
    if ismember(ventas_col_candidates{i}, df.Properties.VariableNames)
        ventas_col = ventas_col_candidates{i};
        break;
    end
end
for i = 1:length(inversion_col_candidates)
    if ismember(inversion_col_candidates{i}, df.Properties.VariableNames)
        inversion_col = inversion_col_candidates{i};
        break;
    end
end

% Debug: show data types and first few values
fprintf('Tipo de datos de x: %s\n', class(df.(inversion_col)));
fprintf('Tipo de datos de y: %s\n', class(df.(ventas_col)));

% Try to convert data, handling different data types
if iscell(df.(inversion_col))
    x = str2double(df.(inversion_col));
else
    x = double(df.(inversion_col));
end

if iscell(df.(ventas_col))
    y = str2double(df.(ventas_col));
else
    y = double(df.(ventas_col));
end

% Check for NaN values and remove them
nan_mask = ~isnan(x) & ~isnan(y);
x = x(nan_mask);
y = y(nan_mask);

n = sum(nan_mask);
fprintf('\nColumnas detectadas -> y: %s, x: %s; n = %d (después de remover NaN)\n', ventas_col, inversion_col, n);

% Debug: show first few values
if n > 0
    fprintf('Primeros valores de x: ');
    fprintf('%.2f ', x(1:min(5,length(x))));
    fprintf('\n');
    fprintf('Primeros valores de y: ');
    fprintf('%.2f ', y(1:min(5,length(y))));
    fprintf('\n');
else
    fprintf('ERROR: No hay datos válidos después de la conversión!\n');
    fprintf('Valores originales de x (primeros 5): ');
    fprintf('%s ', string(df.(inversion_col)(1:min(5,height(df)))));
    fprintf('\n');
    fprintf('Valores originales de y (primeros 5): ');
    fprintf('%s ', string(df.(ventas_col)(1:min(5,height(df)))));
    fprintf('\n');
    return;
end

%%
fprintf('\n(a) Generando diagrama de dispersión y estadísticas descriptivas...\n');
figure;
scatter(x, y);
xlabel('Inversión en redes (millones de pesos)');
ylabel('Ventas (millones de litros)');
title('Dispersión: Ventas vs. Inversión');
grid on;
grid minor;

scatter_path = fullfile(PLOTS_DIR, 'scatter_plot.png');
saveas(gcf, scatter_path, 'png');
fprintf('Figura guardada en: %s\n', scatter_path);


% Calculate statistics for each variable separately
desc_stats = [mean(y), mean(x); 
              median(y), median(x); 
              std(y, 1), std(x, 1); 
              min(y), min(x); 
              quantile(y, 0.25), quantile(x, 0.25); 
              quantile(y, 0.75), quantile(x, 0.75); 
              max(y), max(x)];
desc_table = array2table(desc_stats, 'RowNames', {'media', 'mediana', 'desv_est', ...
                                                  'min', 'Q1', 'Q3', 'max'}, ...
                         'VariableNames', {'Ventas_y', 'Inversion_x'});
fprintf('\nEstadísticas descriptivas:\n');
disp(desc_table);

%%
fprintf('\n(b) Correlación de Pearson e intervalo de confianza al 95%%...\n');
[r, pval] = corr(x, y);
fprintf('r = %.4f\n', r);
fprintf('p-valor (bilateral) = %.6f\n', pval);


z = atanh(r);  
se_z = 1 / sqrt(n - 3);
z_crit = norminv(0.975);  % 1.96
lo_z = z - z_crit * se_z;
hi_z = z + z_crit * se_z;
lo_r = tanh(lo_z);
hi_r = tanh(hi_z);
fprintf('IC 95%%: [%.4f, %.4f]\n', lo_r, hi_r);

%%
fprintf('\n(c) Ajustando RLS (MCO): y = beta0 + beta1*x + e ...\n');

X = [ones(n, 1), x];
[beta, ~, ~, ~, stats] = regress(y, X);
fprintf('\nParámetros estimados:\n');
fprintf('Intercepto: %.4f\n', beta(1));
fprintf('Pendiente: %.4f\n', beta(2));
fprintf('\nResumen del modelo:\n');
fprintf('R^2: %.4f\n', stats(1));
fprintf('F-statistic: %.4f\n', stats(2));
fprintf('p-value: %.6f\n', stats(3));
fprintf('Error estándar: %.4f\n', sqrt(stats(4)));

%%
fprintf('\n(d) Prueba de hipótesis (one-sided, alpha=0.05)...\n');

beta1_hat = beta(2);
se_beta1 = sqrt(stats(4) / sum((x - mean(x)).^2));
df_resid = n - 2;

beta1_H0 = 0.1;

t_stat = (beta1_hat - beta1_H0) / se_beta1;
p_one_sided = 1 - tcdf(t_stat, df_resid);  

fprintf('beta1_hat = %.6f\n', beta1_hat);
fprintf('SE(beta1) = %.6f\n', se_beta1);
fprintf('t = %.4f, df = %d\n', t_stat, df_resid);
fprintf('p-valor (H1: beta1 > 0.1) = %.6f\n', p_one_sided);

alpha = 0.05;
if p_one_sided < alpha
    fprintf('Conclusión: Se RECHAZA H0. La evidencia sugiere un incremento > 50 mil litros por cada 500 mil pesos.\n');
else
    fprintf('Conclusión: NO se rechaza H0 al 5%%. No hay evidencia suficiente para afirmar un incremento > 50 mil litros por cada 500 mil pesos.\n');
end

%%
figure;
scatter(x, y, 'DisplayName', 'Datos');
hold on;
xx = linspace(min(x), max(x), 100);
y_hat = beta(1) + beta(2) * xx;
plot(xx, y_hat, 'DisplayName', 'Recta ajustada');
xlabel('Inversión en redes (millones de pesos)');
ylabel('Ventas (millones de litros)');
title('RLS: Ventas ~ Inversión');
legend('Location', 'best');
grid on;
grid minor;

line_path = fullfile(PLOTS_DIR, 'shampoo_rls_line.png');
saveas(gcf, line_path, 'png');
fprintf('Figura guardada en: %s\n', line_path);

%%
